File: logistics.py
# Every node has a list of child nodes, and a main parent node.
# the first parent of the child gets added to the start of parents list.
# every node has a math value and a display value
# Can choose Trouhgput Limited or Throughput unlimited (TL / TU)
# a node will split until reaches a size of 1

class Node:
  count = 0
  def __init__(self, max_split, max_merge, mathValue, parent = None, needy = False):
    assert max_split >= 2
    assert max_merge >= 2
    assert mathValue >= 1

    Node.count += 1
    self.id = Node.count

    self.parents = [parent]
    self.max_merge = max_merge
    self.max_split = max_split
    self.mathValue = mathValue
    self.children = []
    self.needy = needy
    self.beenSplit = False

  def __str__(self):
    return str(self.id)

  def fillNeedy(self):
    current = self.parents[0]
    while current != None:
      if current.needy:
        self.children.append(current)
        current.parents.append(self)
        current.needy = False
        return True
      current = current.parents[0]
    return False
  
  def split(self):
    if self.mathValue == 1:
      return
    if self.beenSplit:
      return
    self.beenSplit = True
    for i in reversed(range(self.max_split - 1)):
      if self.mathValue % (i + 2) == 0:
        for j in range(i + 2):
          if self.mathValue // (i + 2) == 1:
            if not self.fillNeedy():
              self.children.append(Node(self.max_split, self.max_merge, self.mathValue // (i + 2), self))
          else:
            self.children.append(Node(self.max_split, self.max_merge, self.mathValue // (i + 2), self))
        for child in self.children:
          child.split()
        return
    
    # it is not splitable nicely so we add 1
    self.mathValue += 1
    
    # create new parent
    newParent = Node(self.max_split, self.max_merge, self.mathValue, self.parents[0], True)
    newParent.beenSplit = True
    
    # replace child from parent
    if self.parents[0]:
      for i, child in enumerate(self.parents[0].children):
        if child == self:
          self.parents[0].children[i] = newParent

    newParent.children.append(self)
    self.parents[0] = newParent
        


    for i in reversed(range(self.max_split - 1)):
      if self.mathValue % (i + 2) == 0:
        for j in range(i + 2):
          if self.mathValue // (i + 2) == 1:
            if not self.fillNeedy():
              self.children.append(Node(self.max_split, self.max_merge, self.mathValue // (i + 2), self))
          else:
            self.children.append(Node(self.max_split, self.max_merge, self.mathValue // (i + 2), self))
        for child in self.children:
          child.split()
        return

  def getDigraphLabel(self):
    if self.parents[0] == None:
      shape = 'house'
    
    elif len(self.parents) > 1:
      shape = 'square'
    elif self.mathValue == 1:
      shape = 'invhouse'
    else:
      shape = 'diamond'
      
    return str(self.id) + ' [label="' + str(self.mathValue) + '" shape="' + shape + '"];\n'
  
  def getDigraphConnections(self):
    connections = []
    for child in self.children:
      connections.append(str(self.id) + '->' + str(child.id) + '[label="' + str(self.mathValue // len(self.children)) + '"];\n')
    return connections
  

def createTree(n, maxSplit, maxMerge):
  Node.count = 0
  headNode = Node(maxSplit, maxMerge, n)
  headNode.children.append(Node(maxSplit, maxMerge, n, headNode))
  headNode.children[0].split()
  return headNode





def getAllLabelsAndConnects(node, explored, labels, connections):
  if node in explored:
    return
  explored.append(node)
  labels.append(node.getDigraphLabel())

  for e in node.getDigraphConnections():
    connections += e

  for child in node.children:
    getAllLabelsAndConnects(child, explored, labels, connections)
  
def drawTree(root):
  graphvizString = "digraph G {"
  labels = []
  connections = []
  explored = []

  getAllLabelsAndConnects(root, explored, labels, connections)
  for e in labels:

    graphvizString += e
  for e in connections:
    graphvizString += e
  
  graphvizString += "}"
  print(graphvizString)


  

tree = createTree(11, 3, 3)
drawTree(tree)

File: logistics.js
class TreeNode {
  static count = 0;

  constructor(maxSplit, maxMerge, mathValue, parent = null, needy = false) {
    if (maxSplit < 2 || maxMerge < 2 || mathValue < 1) {
      throw new Error('Invalid input values');
    }

    TreeNode.count += 1;
    this.id = TreeNode.count;

    this.parents = [parent];
    this.maxMerge = maxMerge;
    this.maxSplit = maxSplit;
    this.mathValue = mathValue;
    this.children = [];
    this.needy = needy;
    this.beenSplit = false;
  }

  toString() {
    return String(this.id);
  }

  fillNeedy() {
    let current = this.parents[0];
    while (current !== null) {
      if (current.needy) {
        this.children.push(current);
        current.parents.push(this);
        current.needy = false;
        return true;
      }
      current = current.parents[0];
    }
    return false;
  }

  addChild(value) {
    this.children.push(new TreeNode(this.maxSplit, this.maxMerge, value, this));
  }

  trySplitting() {
    for (let i = this.maxSplit - 2; i >= 0; i--) {
      if (this.mathValue % (i + 2) === 0) {
        for (let j = 0; j < i + 2; j++) {
          if (this.mathValue / (i + 2) === 1) {
            if (!this.fillNeedy()) {
              this.addChild(this.mathValue / (i + 2));
            }
          } else {
            this.addChild(this.mathValue / (i + 2));
          }
        }
        for (const child of this.children) {
          child.split();
        }
        return true;
      }
    }
    return false;
  }

  split() {
    if (this.mathValue === 1) {
      return;
    }
    if (this.beenSplit) {
      return;
    }
    this.beenSplit = true;
    if (this.trySplitting()) {
      return;
    }
    // It is not splittable nicely, so we add 1
    this.mathValue += 1;
    // Create a new parent
    const newParent = new TreeNode(this.maxSplit, this.maxMerge, this.mathValue, this.parents[0], true);
    newParent.beenSplit = true;
    // Replace child from parent
    if (this.parents[0] !== null) {
      for (let i = 0; i < this.parents[0].children.length; i++) {
        if (this.parents[0].children[i] === this) {
          this.parents[0].children[i] = newParent;
        }
      }
    }
    newParent.children.push(this);
    this.parents[0] = newParent;
    if (!this.trySplitting()) {
      throw new Error('Splitting failed');
    }
  }

  getDigraphLabel() {
    let shape = 'house';
    if (this.parents[0] === null) {
      shape = 'house';
    } else if (this.parents.length > 1) {
      shape = 'square';
    } else if (this.mathValue === 1) {
      shape = 'invhouse';
    } else {
      shape = 'diamond';
    }
    return `${this.id} [label="${this.mathValue}" shape="${shape}"];\n`;
  }

  getDigraphConnections() {
    const connections = [];
    for (const child of this.children) {
      connections.push(`${this.id}->${child.id} [label="${this.mathValue / this.children.length}"];\n`);
    }
    return connections;
  }
}

function createTree(n, maxSplit, maxMerge) {
  TreeNode.count = 0;
  const headNode = new TreeNode(maxSplit, maxMerge, n);
  headNode.children.push(new TreeNode(maxSplit, maxMerge, n, headNode));
  headNode.children[0].split();
  return headNode;
}

function getAllLabelsAndConnects(node, explored, labels, connections) {
  if (explored.includes(node)) {
    return;
  }
  explored.push(node);
  labels.push(node.getDigraphLabel());

  for (const e of node.getDigraphConnections()) {
    connections.push(e);
  }

  for (const child of node.children) {
    getAllLabelsAndConnects(child, explored, labels, connections);
  }
}

function drawTree(root) {
  let graphvizString = "digraph G {\n";
  const labels = [];
  const connections = [];
  const explored = [];

  getAllLabelsAndConnects(root, explored, labels, connections);
  for (const e of labels) {
    graphvizString += e;
  }
  for (const e of connections) {
    graphvizString += e;
  }

  graphvizString += "}";

  return graphvizString.replace(/\n/g, '<br>');
}
  
function generateTree(n, maxSplit) {
  tree = createTree(n, maxSplit, 2)
  return drawTree(tree)
}